subprojects/core/src/main/java/org/gradle/configuration/DefaultScriptPluginFactory.java

```diff
@@ -213,7 +213,7 @@ public void apply(final Object target) {
             String operationId = scriptTarget.getId();
             CompileOperation<BuildScriptData> operation = new FactoryBackedCompileOperation<BuildScriptData>(operationId, BODY_COMPILE_STAGE, buildScriptTransformer, buildScriptTransformer, buildScriptDataSerializer);
 
-            final ScriptRunner<? extends BasicScript, BuildScriptData> runner = compiler.compile(scriptType, operation, targetScope.getLocalClassLoader(), ClosureCreationInterceptingVerifier.INSTANCE);
+            final ScriptRunner<? extends BasicScript, BuildScriptData> runner = compiler.compile(scriptType, operation, targetScope, targetScope.getLocalClassLoader(), ClosureCreationInterceptingVerifier.INSTANCE);
```

Comment:
I think this can (and should) actually use `targetScope.getExportClassLoader()` rather than `targetScope.getLocalClassLoader()`, as the target scope represents the scope that contains all of the stuff that should be visible to the script.

Given this and the previous comment, we can just pass in a 'parent' class loader scope to the `compile()` method, rather than passing in a scope and a separate class loader which is always the export classloader of the scope parameter. This will clean up the contract a little, so that `compile()` is given a scope that contains all of the visible things for the script, and the method can create a child scope (or scopes) to hold whatever classes the script container.

Reply:
Makes sense! Done